Make Matrix4.compose more efficient

mqp opened the pull request:
This is a CPU hotspot for rendering if there are a lot of objects in your scene graph. This change makes it moderately more efficient by calculating & writing the resulting matrix values just once, instead of multiple times in the course of applying rotation-scaling-position.

According to my testing, this patch improves overall performance on the `updateMatrixWorld` benchmark in the repository (which just tests calling `updateMatrixWorld` at the root of a large scene graph, like the renderer does) by roughly 10-15% on both Firefox and Chrome.

mqp committed:
Make Matrix4.compose more efficient

mqp committed:
Fix style bugs

WestLangley commented on src/math/Matrix4.js:
```diff
@@ -754,9 +730,38 @@ Object.assign( Matrix4.prototype, {
+		te[ 6 ] = ( yz + wx ) * sy;
+		te[ 10 ] = ( 1 - ( xx + yy ) ) * sz;
+
+		// last column
```
// bottom row

WestLangley commented on src/math/Matrix4.js:
```diff
@@ -754,9 +730,38 @@ Object.assign( Matrix4.prototype, {
+		te[ 7 ] = 0;
+		te[ 11 ] = 0;
+
+		// bottom row
```
// last column

mquander commented:
Is that true? I would have called 3, 7, 11 the last column and 12, 13, 14, 15 the bottom row, the way it is now.

(I didn't write these comments, I just left them where they were, but I'm happy to change them if they really are wrong.)

mquander commented:
Ping -- it would be cool to get this merged.

WestLangley commented:
>Is that true? I would have called 3, 7, 11 the last column

What did you conclude after you checked?

mrdoob commented:
It would also be good if you could share some performance numbers before/after.

mquander commented:
> What did you conclude after you checked?

Ah, OK, I didn't realize the matrices were stored in column-major order (which sounds silly, but I didn't have to think very hard to make this PR, only move work around.) Now I agree that your suggestion is more accurate. I will change those comments.

> It would also be good if you could share some before/after performance numbers.

Sure, is there a particular benchmark which you like to use for measurement? If there is some typical one, I apologize for not knowing about it yet. If not, I will just run the independent `updateMatrixWorld` benchmark in the tests a few more times and output some stats.

mqp committed:
Fix inaccurate comments in Matrix4

These weren't really correct, since Matrix4 elements are stored in
column-major order.

WestLangley commented:
I'd suggest rewriting it like so:
```javascript
compose: function ( position, quaternion, scale ) {

	var te = this.elements;

	var x = quaternion._x, y = quaternion._y, z = quaternion._z, w = quaternion._w;
	var x2 = x + x,	y2 = y + y, z2 = z + z;
	var xx = x * x2, xy = x * y2, xz = x * z2;
	var yy = y * y2, yz = y * z2, zz = z * z2;
	var wx = w * x2, wy = w * y2, wz = w * z2;

	var sx = scale.x, sy = scale.y, sz = scale.z;

	te[ 0 ] = ( 1 - ( yy + zz ) ) * sx;
	te[ 1 ] = ( xy + wz ) * sx;
	te[ 2 ] = ( xz - wy ) * sx;
	te[ 3 ] = 0;

	te[ 4 ] = ( xy - wz ) * sy;
	te[ 5 ] = ( 1 - ( xx + zz ) ) * sy;
	te[ 6 ] = ( yz + wx ) * sy;
	te[ 7 ] = 0;

	te[ 8 ] = ( xz + wy ) * sz;
	te[ 9 ] = ( yz - wx ) * sz;
	te[ 10 ] = ( 1 - ( xx + yy ) ) * sz;
	te[ 11 ] = 0;

	te[ 12 ] = position.x;
	te[ 13 ] = position.y;
	te[ 14 ] = position.z;
	te[ 15 ] = 1;

	return this;

},
```
And of course, confirming the returned value is correct.

mquander commented:
Sure, although analogous code in a couple other functions in `Matrix4` are written with assignments in the row-by-row style (`multiplyMatrices`, `makeRotationFromEuler`) as well. Let me know if you prefer me to fix those up too.

mqp committed:
Prefer linear assignment order in Matrix4.compose

mquander commented:
[Here's some results from the `updateMatrixWorld` benchmark.](https://docs.google.com/spreadsheets/d/1uXJP4c8pekhgztsC5eLH0L1JgebTfPLLlcmZB4IZ7oE/edit?usp=sharing)

I don't know whether the differences in performance between different scene graph shapes are really signifying something or whether they're just an artifact of the fairly unscientific methodology. The improvement seems unambiguous, though.

WestLangley commented:
> Let me know if you prefer me to fix those up too.

I would only change them if there is a compelling reason to do so.  :)

mrdoob commented:
@takahirox any chance you could do some performance test comparing the old and new `compose()` methods?

takahirox commented:
A bit off topic. Wondering if introducing `.matrixNeedsUpdate` would be worth for the performance.

The idea is, set `onChange` to `Object3D`'s `position/quaternion/scale` in which set `.matrixNeedsUpdate` true. `Object3D` calls `.updateMatrix()` in `.updateMatrixWorld()` only when `.matrixNeedsUpdate` is `true`. In short, we don't call `.updateMatrix()` if we don't need to.

I applied quick patch and confirmed 15-20 % better number on `updateMatrixWorld` benchmark.

If you folks are interested in, I'll make PR and we can discuss.

mrdoob commented:
@takahirox `Euler` and `Quaternion` already have a `onChange` code on them to keep them in sync and it's a nightmare to maintain... 😣

takahirox commented:
Yup, that's the disadvantage of this idea. But improving the hotspot performance 15-20 % isn't bad. So, I'll make PR so far, we can see how easy/hard to maintain.

WestLangley commented:
@mrdoob @takahirox BTW, awhile ago, I hacked the Euler code so it never synchronizes with the quaternion until the euler is accessed, and then it synchronizes all the time -- sort of like lazy updating, but much simpler conceptually. 

Unfortunately, I saw no measurable performance improvement.

mquander commented:
If a `matrixNeedsUpdate` flag was implemented, it seems to me that you could then additionally say that for a given object C, if neither C nor any of C's ancestors had `matrixNeedsUpdate`, then they must have all stayed put in world-coordinates too and therefore you don't need to call `multiplyMatrices` to recompute `matrixWorld` for C. That would be an additional benefit on top of saving the `compose` work.

WestLangley commented:
@takahirox I you are are going to pursue lazy updating of matrices or Euler, then be sure to check the history of this topic. There has been a lot of work in this regard, especially by @gero3.
